### ISSUE TITLE
`magicgui` doesn't return `Layer` if return type is `Optional`

### ISSUE THREAD

**Issue by Modjular** [napari/napari#7313]
### 🐛 Bug Report

I expected my widget to return a new `Image` layer.
It did not.

### 💡 Steps to Reproduce

Wrap the return type of your widget in `Optional[]`. (Modern format works too: `Image | None`)

```python
import napari
import numpy as np
from napari.layers import Image
from typing import Optional

@magicgui
def my_widget() -> Optional[Image]:
    return Image(np.random.rand(300, 300), name='my Image')

viewer = napari.Viewer()
viewer.window.add_dock_widget(my_widget)
my_widget()

# ... will fail to produce a new layer in the viewer
```

### 💡 Expected Behavior

I expect a new layer to appear in the viewer

### 🌎 Environment

napari: 0.5.4
Platform: macOS-12.6-arm64-arm-64bit
System: MacOS 12.6
Python: 3.10.14 (main, Mar 20 2024, 03:57:45) [Clang 14.0.0 (clang-1400.0.29.202)]
Qt: 6.7.1
PyQt6: 
NumPy: 1.26.3
SciPy: 1.11.4
Dask: 2024.9.1
VisPy: 0.14.3
magicgui: 0.9.1
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.3.0
psygnal: 0.11.1
npe2: 0.7.7
pydantic: 2.5.3

OpenGL:
- GL version: 2.1 Metal - 76.3
- MAX_TEXTURE_SIZE: 16384
- GL_MAX_3D_TEXTURE_SIZE: 2048

Screens:
- screen 1: resolution 1512x982, scale 2.0
- screen 2: resolution 5120x1440, scale 1.0

Optional:
- numba not installed
- triangle not installed
- napari-plugin-manager not installed

Settings path:
- /Users/tony/Library/Application Support/napari/.venv_b6d25b91631b21575809fb71a632dc1edf1f08b3/settings.yaml

### 💡 Additional Context

Let me know if this the expected behavior, or if there's an explanation for why `Optional[]` is interpreted this way.

**Comment by Czaki** [comment]
It is definitely bug.

**Comment by Modjular** [comment]
@Czaki if you can point me in the right direction I'll take a stab at fixing it.

**Comment by Czaki** [comment]
We regiser it in this line:

https://github.com/napari/napari/blob/baaf01dadcdef62bdb9a14e8a1df4ca974c84a49/napari/layers/base/base.py#L119

where we call the function from magicgui. 

@tlambert03 It looks like the next place to fix handling Optional in magicgui. Did I'm correct?

**Comment by tlambert03** [comment]
> @tlambert03 It looks like the next place to fix handling Optional in magicgui. Did I'm correct?

It could be fixed in one of two places. 

1. You could fix your callback here in napari, to:
    ```python
    register_type(Optional[Layer], return_callback=add_layer_to_viewer)
    ```
    and then make sure that `add_layer_to_viewer` does the right thing if it receives an input of `None`.

1. magicgui could be modified to do a little additional logic to make it such that, even if `register_type` was invoked with the non-optional form of the type as napari has done (`register_type(Layer, return_callback=...)`), then if a magicgui widget states that it might return `None | Layer`, we should still call that function anyway in case it doesn't return `None`, and just discard the result if it is None.

The first option is a bit more explicit in terms of typing, but I'm also not opposed to the second option if you'd like that to be the way this is solved.  In which case you can transfer this issue to magicgui if you'd like

**Comment by tlambert03** [comment]
more broadly speaking, i think while the second option does seem straightforwardly elegant in this specific case, how well does it generalize to other unions.  For example if someone registers a FunctionGui that returns `int | Layer | None | str` ... should we still call it, check the output type, and then only call napari's callback if it's a Layer?  That gets into the weeds of having to decide whether a runtime object is compatible with a type annotation (which I learned is much hairier than I thought when I worked on the DOOR stuff in https://beartype.readthedocs.io/en/latest/api_door/).

so, if napari registers `Optional[Layer]`, then we can continue living blithely in our simple world of direct type mapping.  while the second option edges (quite reasonably) over what could turn into a slippery slope

**Comment by DragaDoncila** [comment]
Hmm yeah I'm leaning towards option 1 here. napari should explicitly register `Optional[Layer]` and it's up to our callback to make sure it ignores the `None`. I agree that moving this functionality to magicgui semi-defeats the purpose of the type annotation in the first place, if (in the extreme case) we're just always running callbacks on the off chance they return the thing we registered a processor for...

**Comment by jni** [comment]
I'm 👍 on fixing it on napari's side.

As an aside, apologies if this is obvious for those more familiar with typing, but I thought this was cool!

```python
In [1]: from typing import Optional

In [2]: Optional[int] == int | None
Out[2]: True
```

**Comment by Czaki** [comment]
@Modjular So it looks like the best solution will be to change registration from decorator on Layer.

Personally, I prefer to move it to `napari.types::_register_types_with_magicgui` to remove "on import" effect (we should also call `_register_types_with_magicgui` more explicitly). 

@jni @DragaDoncila What did you think about the proper place to call  `register_type`?

**Comment by DragaDoncila** [comment]
@Czaki I agree, let's get rid of the decorator and move it to `napari.types`. I also agree it's weird to just call it at the end of that file... perhaps somewhere early in the viewer constructor?
